CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc

```diff
@@ -384,180 +416,262 @@ void HcaluLUTTPGCoder::update(const HcalDbService& conditions) {
+      const HcalQIEShape* shape = conditions.getHcalShape(cell);
+      const HcalPedestal* effPedestals = conditions.getEffectivePedestal(cell);
+      HcalCoderDb coder(*channelCoder, *shape);
+      const HcalLutMetadatum* meta = metadata->getValues(cell);
+
+      auto tpParam = conditions.getHcalTPChannelParameter(cell, false);
+      int weight = tpParam->getauxi1();
+
+      int lutId = getLUTId(cell);
+      int lutId_ootpu = lutId + sizeZDC_;
+      Lut& lut = inputLUT_[lutId];
+      Lut& lut_ootpu = inputLUT_[lutId_ootpu];
+      float ped = 0;
+      float gain = 0;
+      float pedWidth = 0;
+      uint32_t status = 0;
+
+      if (LUTGenerationMode_) {
+        const HcalCalibrations& calibrations = conditions.getHcalCalibrations(cell);
+        for (auto capId : {0, 1, 2, 3}) {
+          ped += calibrations.effpedestal(capId);
+          gain += calibrations.LUTrespcorrgain(capId);
+          pedWidth += effPedestals->getWidth(capId);
+        }
+        ped /= 4.0;
+        gain /= 4.0;
+        pedWidth /= 4.0;
+
+        const HcalChannelStatus* channelStatus = conditions.getHcalChannelStatus(cell);
+        status = channelStatus->getValue();
+
+      } else {
+        const HcalL1TriggerObject* myL1TObj = conditions.getHcalL1TriggerObject(cell);
+        ped = myL1TObj->getPedestal();
+        gain = myL1TObj->getRespGain();
+        status = myL1TObj->getFlag();
+      }  // LUTGenerationMode_
+
+      ped_[lutId] = ped;
+      gain_[lutId] = gain;
+      bool isMasked = ((status & bitToMask_) > 0);
+      float rcalib = meta->getRCalib();
+
+      auto adc2fC = [channelCoder, shape](unsigned int adc) {
+        float fC = 0;
+        for (auto capId : {0, 1, 2, 3})
+          fC += channelCoder->charge(*shape, adc, capId);
+        return fC / 4;
+      };
+
+      int qieType = conditions.getHcalQIEType(cell)->getValue();
+
+      const size_t SIZE = qieType == QIE8 ? INPUT_LUT_SIZE : UPGRADE_LUT_SIZE;
+      const int MASK = QIE10_ZDC_LUT_BITMASK;
+
+      lut.resize(SIZE, 0);
+      lut_ootpu.resize(SIZE, 0);
+
       for (unsigned int adc = 0; adc < SIZE; ++adc) {
-        if (isMasked)
+        if (isMasked) {
           lut[adc] = 0;
-        else {
-          lut[adc] =
-              std::min(std::max(0, int((adc2fC(adc) - ped) * gain * rcalib / lsb_ / cosh_ieta_[cell.ietaAbs()])), MASK);
-          if (adc > FG_HF_thresholds_[0])
-            lut[adc] |= QIE10_LUT_MSB0;
-          if (adc > FG_HF_thresholds_[1])
-            lut[adc] |= QIE10_LUT_MSB1;
+          lut_ootpu[adc] = 0;
+        } else {
+          if ((adc2fC(adc) - ped) < (pedWidth * 5.)) {
+            lut[adc] = 0;
+            lut_ootpu[adc] = 0;
+          } else {
+            lut[adc] = std::min(std::max(0, int((adc2fC(adc) - ped) * gain * rcalib / zdc_lsb_)), MASK);
+            lut_ootpu[adc] =
+                std::min(std::max(0, int((adc2fC(adc) - ped) * gain * rcalib * weight / (zdc_lsb_ * 256))), MASK);
```

Comment:
What is this "256"? Anything to do with `UPGRADE_LUT_SIZE` (or maybe `SIZE`)?

Reply:
The 256 is because the "weight" is encoded as an 8-bit integer, where in the hardware the true weight applied = weight / (# of 8 bit values) = weight / 256. 

So this does not have to do with UPGRADE_LUT_SIZE, they just happen to have the same value. 

I can define a new constant to make the reason for this 256 more clear, something like EIGHT_BIT_VARIABLE_SIZE